keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1210,230 +1099,251 @@ server-side-apply (SSA) simpler. Patching individual list elements is not
+// and ignore all others.
+//
+// If necessary, a consumer can check the number of total devices in a pool (included
+// in each slice) to determine whether its view of a pool is complete.
+//
+// For devices that are not local to a node, the node name is not set. Instead,
+// the driver may use a node selector to specify where the devices are available.
 type ResourceSlice struct {
     metav1.TypeMeta
     // Standard object metadata
+    // +optional
     metav1.ObjectMeta
 
-    // NodeName identifies the node which provides the resources
-    // if they are local to a node.
-    //
-    // A field selector can be used to list only ResourceSlice
-    // objects with a certain node name.
-    NodeName string
+    Spec ResourceSliceSpec
+
+    Status ResourceSliceStatus
+}
 
+type ResourceSliceSpec struct {
     // DriverName identifies the DRA driver providing the capacity information.
     // A field selector can be used to list only ResourceSlice
     // objects with a certain driver name.
+    //
+    // Must be a DNS subdomain and should end with a DNS domain owned by the
+    // vendor of the driver.
     DriverName string
 
-    ResourceModel
+    // PoolName is used to identify devices. For node-local devices, this
+    // is often the node name, but this is not required.
+    //
+    // It must not be longer than 253 and must consist of one or more DNS sub-domains
+    // separated by slashes.
+    PoolName string
+
+    // NodeName identifies the node which provides the devices.
+    // A field selector can be used to list only ResourceSlice
+    // objects belonging to a certain node.
+    //
+    // This field can be used to limit access from nodes to slices with
+    // the same node name. It also indicates to autoscalers that adding
+    // new nodes of the same type as some old node might also make new
+    // devices available.
+    //
+    // NodeName and NodeSelector are mutually exclusive.
+    // If both are unset, then the devices are available in the
+    // entire cluster.
+    //
+    // +optional
+    NodeName string
+
+    // Defines which nodes have access to the devices in the pool.
+    // Must not be set when the node name is set.
+    //
+    // NodeName and NodeSelector are mutually exclusive.
+    // If both are unset, then the devices are available in the
+    // entire cluster.
+    //
+    // +optional
+    NodeSelector *v1.NodeSelector
+
+    // The generation gets bumped in all slices of a pool whenever device
+    // definitions change. A consumer must only use device definitions from slices
+    // with the highest generation number and ignore all others.
+    PoolGeneration int64
+
+    // The total number of devices in the pool.
+    // Consumers can use this to check whether they have
+    // seen all slices.
+    PoolDeviceCount int64
+
+    // SharedCapacity defines the set of shared capacity consumable by
+    // devices in this ResourceSlice.
+    //
+    // Must not have more than 128 entries.
```

Comment:
I don't necessarily share this concern. I don't think it will actually be a problem in practice.

As a vendor, I **really** like the ability to reason separately about any possible `SharedCapacity`and the concrete set of `Devices` I want the scheduler to be able to allocate on my behalf. 

This is my mental model as a vendor:

1. Simple devices are nothing more than a named list of attributes. I create an array of strings to represent each device and I attach some attributes to each of them. Users can then select on these attributes to help narrow down the scheduler's choice of which device to pick.  

1. Complex devices can have both attributes and some number of quantifiable resources attached to them (what we've been calling `capacity`). These quantifiable resources are either local to a device, or pulled from some shared pool. Just as before, I enumerate the full set of devices I want the scheduler to be able to allocate and attach some attributes to them. If the devices have some local "capacity" they want to advertise, I add this too. If a device needs to pull some "capacity" from a shared pool, I make sure there is an entry in the top level `SharedCapacity` that defines the "limit" of this shared resources, and then advertise how much of this shared `capacity` will be made available to users if the scheduler decides to pick this device.

Reply:
You don't share which concern?  It seems inevitable to be a problem.  If I have 8 GPUs and each GPU has 16 shared items (mutexes), that's the limit.  So that forces me to organize my slices dependant upon how many shared items I have.  If I have 16 memory blocks (mutex style) and 1 other counted shared resource, then I can only put 7 devices in a slice. But looking at the dgxa100 example, there are 26 "devices" per GPU, so actually I can only have 4 such devices in a slice.

Will your driver maintainers eventually just say "one card per slice is easier" and leave it at that?  IF we think that will be the common usage mode, can we lean into it and get a simpler API?  Remember many parties will have to implement this API, so simpler is a feature.

Reply:
I think we're making progress on this in https://github.com/kubernetes-sigs/wg-device-management/issues/20 - but I don't think it's going fast enough to make 1.31.